### DISCUSSION TITLE
What are you all working on at the moment?

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#204] [General]
Just curious about what little ctrlr rabbit holes you are all up to your necks in.

This is what I'm busily trying to squash bugs on, at the moment. After this is done, I'm going to work on Kurzweil panels with bijlevel.

https://godlike.com.au/index.php?id=150

**Comment by dnaldoog** [comment]
A panel for the Korg MS2000!

**Reply by Godlike-Productions** [reply]
Oooooo, nice.  I have one at home, and it was on my todo list so I could remote control it while I'm on site.  Let me know if you need any testing done.

**Reply by dnaldoog** [reply]
Excellent! I was hoping I could find someone to test it. It seems to be very complex. Am in the early stages.

**Reply by inteyes** [reply]
I am waiting MS2000 to arrive in about a month or so, I was thinking of making my own panel / librarian for it, and it's good to hear that somebody already working on this. Would be great to test it if there will be a chance :)

**Comment by tomlem** [comment]
A panel for the Roland U-220.

**Comment by Godlike-Productions** [comment]
Jupiter 8 Morph has moved into Beta.  Going to try to get an Easter Egg Feature in over the next couple of day. 

Copy Protection and final builds on Tuesday then time for the Beta Testers to break it.

What an effort, but I've learnt so much.

![Config Beta 00](https://user-images.githubusercontent.com/13851825/110232970-1cb0ce00-7f5c-11eb-91e5-549006a6b4a9.png)
![Panel Beta 00](https://user-images.githubusercontent.com/13851825/110232972-1e7a9180-7f5c-11eb-9077-3867a028c6cf.png)
![Patch Beta 00](https://user-images.githubusercontent.com/13851825/110232974-1f132800-7f5c-11eb-8a31-e63c605c0dfb.png)

**Reply by bijlevel** [reply]
Congratulations! Looks very, very impressive!

**Comment by bijlevel** [comment]
On my todo list is the Proteus/1 from E-mu Systems Inc. On the old forum is a panel for the Proteus/2 which uses essentially the same sysex, but that one is still in its "kinderschoenen" (children's shoes) as we from The Netherlands say, when something is far from finished. The sysex format is very well documented (and I have the manual) and it does not seem very complicated, but a lot of work for the UI, lots of knobs and sliders..

**Comment by DRBAUS** [comment]
I am still working on a Panel for the Ensoniq Mirage running the alternative Soundprocess OS but I am running into some issues with Ctrlr that I cannot pinpoint and I am getting frustrated now. I will start a topic for it.

**Comment by Ted-juh** [comment]
I'm working on small side-projects that I hope to implement in the panel I'm making for the Roland Sh-201 (or its newer equivalent the Roland Gaia). The "min/ max slider", posted about that in the discussions, is now nearly completed and functional. Also working on an arpeggiator that's fully GUI, with adjustable note-lengths and velocity. And working on a re-design of the panel, just because I can. And to make matters worse, I want to make a component that lets me save voices independent from the "bank manager" of the synth. And a "Multi-Slider" component that lets me adjust multiple sliders at the same time. Hence the "min/ max slider", which can be set to certain limits regardless of the "Multi-Slider" that goes full range.

And mostly thanks to Dnaldoog and other fellows like Goodweather who gave valuable input/ insights on how to program certain things, it's all coming along nicely. In the background, there's a lot of research about how Juce works or what you can do with Lua (create your own classes). The conjunction of those two factors goes deep, very deep. But those insights let you write complete new functions that work better or more intuitively or are faster than the ones you had. 

So, what am I working at? Short reply: Too much..

**Comment by mtarenskeen** [comment]
I have been working on an update of my panel for the Reface DX, after an exciting discovery of an undocumented SysEx feature: It always looked as if the Reface DX could only receive and send one single patch to/from the temporary editbuffer. This was a very disappointing limitation compared to the old classic Yamaha FM synths (DX7, DX100, DX21, T81Z) this Reface DX was supposed to re-create. With the new SysEx info I was able to make bulkdumps of all 32 patches as a whole.  Available from http://ctrlr.martintarenskeen.nl

**Reply by Ted-juh** [reply]
How does one find undocumented sysex? Is it a simple manner of just sending 0-127 values to a different address and hear what it does to the sound of a synth or just look what the synth replies?

**Reply by mtarenskeen** [reply]
In many case it's more a question of incomplete, insufficient, or erratic sysex documentation. For example for my Korg DS-8 and 707 panels. The manual did show me SysEx commands for a dump request, and the SysEx headers in the dumps. But it did not specify where each parameter was situated in the dump. It did show a command for parameter changes, but there was not a list with all editable parameters and addresses. I have done a lot of trial and error, changing each individual parameter, make a dump, examining the dump in a hex editor and/or SysEx dump utility, and thus identifying the correct address for each parameter.

It also helps if you examine the SysEx documentation of other products of the same manufacturer. Each manufacturer has it's own housestyle for SysEx implementations. From the beginning, when manufacturers agreed about the MIDI Specification, System Exclusive data were meant to give manufacturers the freedom to add their own unique features to their products without breaking the MIDI standards. This makes it impossible to give a general answer to your question.

My latest update of the Reface DX Ctrlr panel uses undocumented SysEx dump requests. Someone experimented with the documented request messages, and wanted to know if changing one or two of the bytes in the SysEx string would give any usable reponse from the Reface DX. More or less by accident he found two working values (out of 128). One was the documented value, but the other was giving mysterious responses and he mailed me about it. After more experiments this has lead to very useful additional information that I used to improve my Reface DX panel.

**Reply by Godlike-Productions** [reply]
For Kurzweil products, we've been reverse engineering the file and dump formats for over 10 years.  We also snoop MIDI traffic from existing editors (even broken ones) and then lots  and lots of trial and error.

**Reply by bijlevel** [reply]
Isn't it typical ( or a shame?) that one needs (in some cases 10) years to find out how the sysex of a midi instrument is implemented, while there are midi editors like MidiQuest and Unisyn out there of which the developers obvious had/have access to the sysex implementation? I'm afraid that for instance in the case of Kurzweil that company has/had some contractual agreement with those editor developers. I did write to Kurzweil about this, but they never answered.
For my remaining instruments, e.g. Proteus 1 and Alesis DM5 I'm lucky they have the sysex well documented, as was the case with the Yamaha DX 7/TX 7. It is already a PITA (and therefore time consuming) to understand even the documentation, let alone the programming in Ctrlr! But I comfort myself with the idea that developing Ctrlr panels is a hobby ;-).

**Reply by Godlike-Productions** [reply]
The developers would likely have had access to some of Kurzweil's codebase, and access to one of the tech team who likely is no longer there.  They never documented file format in easily digestible formats. I have an NDA with them, so can get more stuff than most people, but they literally have to go diving into code for products that they no longer support or are developing to get answers, and that can take time.

One of the issues with the Kurzweil's is that the number of parameters is immense.  Thousands of them, and because the architecture isn't fixed, there is an added difficulty in building true editors.

**Comment by mtarenskeen** [comment]
Long before Ctrlr existed, in the early days of MIDI, I have created many so called "device adaptors" for an Atari ST program named UNI_MAN. I learnt a lot about weird SysEx implementations used by different manufacturers in those days. For the real vintage lovers UNI_MAN and my adaptors are still available from my website http://uniman.martintarenskeen.nl
